hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java

```diff
@@ -393,18 +396,28 @@ public HoodieArchivedTimeline getArchivedTimeline(String startTs) {
 
   /**
    * Validate table properties.
-   * @param properties Properties from writeConfig.
+   * @param writeConfigProps Properties from writeConfig.
    */
-  public void validateTableProperties(Properties properties) {
+  public void validateTableProperties(Properties writeConfigProps) {
+    // Once table is configured to be append-only, it cannot be mutable or allow setting record key or precombine fields for updates
+    if (getTableConfig().isAppendOnlyTable()) {
+      boolean appendOnlyTable = Boolean.parseBoolean(
+          writeConfigProps.getProperty(HoodieTableConfig.APPEND_ONLY_TABLE.key(), String.valueOf(HoodieTableConfig.APPEND_ONLY_TABLE.defaultValue())));
+      checkArgument(appendOnlyTable, String.format("%s is enabled. Please recreate the table with record key to support mutable tables.", HoodieTableConfig.APPEND_ONLY_TABLE.key()));
+      checkState(isNullOrEmpty(writeConfigProps.getProperty(HoodieTableConfig.RECORDKEY_FIELDS.key())), String.format("%s set for an append-only table", HoodieTableConfig.RECORDKEY_FIELDS.key()));
+      checkState(isNullOrEmpty(writeConfigProps.getProperty("hoodie.datasource.write.recordkey.field")), "hoodie.datasource.write.recordkey.field set for an append-only table");
```

Comment:
These are leaking Spark impl details in here. We should vet these out on different level